yarpl/include/yarpl/Refcounted.h

```diff
@@ -318,14 +318,32 @@ Reference<CastTo> make_ref(Args&&... args) {
   );
 }
 
-template <typename T>
-Reference<T> get_ref(T* object) {
-  static_assert(
-      std::is_base_of<Refcounted, std::decay_t<T>>::value,
-      "Reference can only be constructed with a Refcounted object");
+class enable_get_ref {
+ private:
+  // force the class to be polymorphic so we can dynamic_cast<T>(this)
+  virtual void dummy_internal_get_ref() {}
+
+ protected:
+  // materialize a reference to 'this', but a type even further derived from
+  // Derived, because C++ doesn't have covariant return types on methods
+  template <typename As>
+  Reference<As> ref_from_this(As* ptr) {
```

Comment:
do we need overload with "const As*" ? so we can call ref_from_this in a const method?

Reply:
Also a good point; we don't really have many uses of 'const' in the codebase, but that's not exactly something to strive towards maintaining. I'll add a 'const' version.